### PR TITLE
ThreadManager: StealAndDropActiveLocks in the child forked process

### PR DESCRIPTION
**Pull request by Sonicadvance1**
The creation mutex could have been held if the parent thread was in the
middle of creating a thread when forking. This would result in a
deadlock once the fork child attempted to create another thread.

Forcefully dropping the lock in the fork child works around this
deadlock. This comes at the expense of potentially leaving resources
guarded by the thread creation mutex in an invalid state. Crashes caused
by this are easier to reason about than a delayed deadlock, though.

### CODE REVIEW

**Review by neobrain — COMMENTED** [review]
Could you provide more context to this commit message?

> While this doesn't fix the deadlock issue, it exchanges a hard to reason
about deadlock with a potential crash which is significantly better than
before.

From the git log it won't be clear what deadlock issue you're referring to. It's also unclear how you're turning a deadlock into a potential crash.

File: Source/Tools/LinuxEmulation/LinuxSyscalls/ThreadManager.cpp [hunk 1]
```diff
@@ -185,8 +185,6 @@ namespace FEX::HLE {
   }
 
   void ThreadManager::UnlockAfterFork(FEXCore::Core::InternalThreadState *LiveThread, bool Child) {
```
[the remark this answers is not in the document]
Comment (Sonicadvance1) [reply Source/Tools/LinuxEmulation/LinuxSyscalls/ThreadManager.cpp:187]: I'd claim the added line of `ThreadCreationMutex.StealAndDropActiveLocks();` is doing something lock related.
There will also be additional work in this function in the following months which does more with that mutex.

No need to rename to a new random function name and then rename it back later.

[the remark this answers is not in the document]
Comment (neobrain) [reply Source/Tools/LinuxEmulation/LinuxSyscalls/ThreadManager.cpp:187]: Fair. Not sure this fits into UnlockAfterFork either, but sounds like it'll be easier to come up with a more fitting name once the coming refactors drop in.

**Comment by Sonicadvance1** [comment]
> Could you provide more context to this commit message?
> 
> > While this doesn't fix the deadlock issue, it exchanges a hard to reason
> > about deadlock with a potential crash which is significantly better than
> > before.
> 
> From the git log it won't be clear what deadlock issue you're referring to. It's also unclear how you're turning a deadlock into a potential crash.

When a fork occurs there are a handful of mutexes in-play that can result in a deadlock between multiple of them.
- Thread Creation mutex
- VMA allocator mutex
- Code Cache Invalidation mutex
- (maybe) per-thread Lookup cache invalidation mutex.

During a fork we currently lock the code cache invalidation mutex and the VMA allocator mutex.
This leaves the thread creation mutex in an unlocked state.
Once the fork occurs this thread creation mutex is in an unknown state on the child process.

The obvious solution would be to lock the thread creation mutex before forking so it is in a known state. The problem with this is that the VMA allocator mutex, the code cache invalidation mutex, and the thread creation mutex all interconnect with each other in hard to follow patterns. Locking beforehand has basically a 100% chance of hardlocking all processes as they fork. This will be rewritten going forward once I detangle the details as for how the implementation is. 

So the current workaround is to leave the Thread creation mutex in whichever state it is going in to the fork, and then clearing it to the default initialization value. Since there is only one thread remaining this is "safe" since we know the VMA mutex and code cache invalidation mutexes were locked so nothing could be touching memory mutexes or code cache. The potential for crashing comes down to some of the weird states that the newly forked process could be in if another thread was in the process of creating a thread (instead of a fork!) where it has only locked the thread creation mutex. Likely to result in a crash if say, the fextl::vector of thread objects was in the middle of getting modified when the fork occured.

**Comment by neobrain** [comment]
Thanks, that's a great explanation!

The wording that I pointed to actually gave me the wrong idea about what's happening (thinking you were referencing some ominous "deadlock issue" that was unrelated). If I understand correctly, "the deadlock issue" is just referencing the behavior described in the first paragraph. Is that correct? It's sadly not at all obvious that the first paragraph describes a deadlock.

The following wording makes this clearer, assuming my above interpretation is correct:
```
The creation mutex could have been held if the parent thread was in the middle of creating a thread when forking. This would result in a deadlock once the fork child attempted to create another thread.

Forcefully dropping the lock in the fork child works around this deadlock. This comes at the expense of potentially leaving resources guarded by the thread creation mutex in an invalid state. Crashes caused by this are easier to reason about than a delayed deadlock, though.
```

Not sure what to do with the addendum `This is the correct approach here, but we may still need to shuffle
around some mutex usage.`. This kind of reads as "This added line will probably stay, but maybe not", which isn't very informative.

**Comment by Sonicadvance1** [comment]
> Thanks, that's a great explanation!
> 
> The wording that I pointed to actually gave me the wrong idea about what's happening (thinking you were referencing some ominous "deadlock issue" that was unrelated). If I understand correctly, "the deadlock issue" is just referencing the behavior described in the first paragraph. Is that correct? It's sadly not at all obvious that the first paragraph describes a deadlock.

Yep.

> 
> The following wording makes this clearer, assuming my above interpretation is correct:
> 
> ```
> The creation mutex could have been held if the parent thread was in the middle of creating a thread when forking. This would result in a deadlock once the fork child attempted to create another thread.
> 
> Forcefully dropping the lock in the fork child works around this deadlock. This comes at the expense of potentially leaving resources guarded by the thread creation mutex in an invalid state. Crashes caused by this are easier to reason about than a delayed deadlock, though.
> ```

Sure.

> 
> Not sure what to do with the addendum `This is the correct approach here, but we may still need to shuffle around some mutex usage.`. This kind of reads as "This added line will probably stay, but maybe not", which isn't very informative.

Yep.

Replaced the commit message with yours.

I really can't give a shit about the minutiae of detail.

**Comment by neobrain** [comment]
Nodding off the PR itself since the misleading commit message has been fixed. I'm intrigued to see what a cleaner approach to synchronization here will look like.

> I really can't give a shit about the minutiae of detail.

This comment has been discussed internally.

**Review by neobrain — APPROVED** [review]
